### ISSUE TITLE
Avoid duplicative error printing in ErrorHandler

### ISSUE THREAD

**Issue by alexzorin** [certbot/certbot#8320]
This issue is about double-printing of exceptions that are raised by the recovery function of `ErrorHandler`.

The behavior is explained in https://github.com/certbot/certbot/pull/8310#issuecomment-697046235.

---

> What if we did the same thing by putting a log=True default parameter into the constructor of ErrorHandler? That would allow control of logging at the actual site that is being protected, in a generic way. Being an internal module, it should be safe?

Would that be different from just not having one of the locations print an error? The nginx plugin just raises an error, it doesn't call `ErrorHandler` directly.

_Originally posted by @ohemorange in https://github.com/certbot/certbot/pull/8310#issuecomment-698586246_

**Comment by alexzorin** [comment]
>Would that be different from just not having one of the locations print an error?

I guess it would be doing exactly that. 

For my own clarified understanding, here is the detail of what I believe is happening in [the gist](https://gist.github.com/alexzorin/a0ce1781ee6aa5f3bd8ba0595008cf05), which is quite a very contrived scenario but exercises the error handling quite well:

1. I am holding port 80 hostage with an external process, and nginx is stopped.
2. I ask Certbot to install a certificate using the nginx plugin. It enters `certbot.client.deploy_certificate`.
3. Certbot modifies the nginx config and saves it. This succeeds.
4. Certbot goes to restart nginx after installing the certificate. This is protected by an `ErrorHandler`. The restart fails.
5. `ErrorHandler` logs the error at `debug` level only.
6. `ErrorHandler` calls the recovery function (`certbot.client._rollback_and_restart`). 
7. The recovery function fails, and `ErrorHandler` logs the exception at the `error` level.
8. Because `ErrorHandler.__exit__` returns `False` for the exception raised by the recovery function, the exception continues up the call stack.
9. The global handler eventually catches it and prints it again.

----

For the terminal output, I think it would be helpful to:

1. Inform the user that the certificate installation failed.
2. Inform the user that the configuration rollback failed as well.
3. Do (1) and (2) only one time each.

I am not actually sure whether we actually want to show the excessive nginx output in either case, what do you think?

My idea was to pass in `log=False` to `ErrorHandler` from `certbot.client.deploy_certificate`, suppressing the `Encountered exception during recovery` output. Removing terminal logging in (7), leaving us with only one instance of output in (9). We can prettify (9) as required.

>At a high level, the problem is that the same restart function is getting called for its own recovery routine. We could possibly fix this by subclassing `MisconfigurationError` as something like `RestartError`, and testing for that in `ErrorHandler`. What do you think about that?
 
I think this has gone completely over my head, sorry. 

Yes, `installer.restart` gets called twice. This seems appropriate to me. Once trying to install the certificate, once rolling it back.

The output is not appropriate, as we see the error output from the rolling back operation twice.

Could you ELI5 your approach?

**Comment by ohemorange** [comment]
> Yes, installer.restart gets called twice. This seems appropriate to me. Once trying to install the certificate, once rolling it back.

Thinking it through more, this would probably take more work than I initially imagined. Basically it depends on being able to differentiate a restart error due to a misconfiguration (what we assume in the code basically, since we assume that we just messed it up, and where a rollback and restart is a good option to take) vs. a restart error due to the port being in use. If we could detect the latter, then raise a specific error for it, we would know that there's no point trying to do a restart as part of the rollback and restart, since restarting *just* failed during install. This check would probably be best in `_rollback_and_restart`, rather than `ErrorHandler` as I erroneously stated above. But relying on parsing the nginx output is maybe actually more than we want to be doing here.

> For the terminal output, I think it would be helpful to:
>
> 1. Inform the user that the certificate installation failed.
> 2. Inform the user that the configuration rollback failed as well.
> 3. Do (1) and (2) only one time each.
>
> I am not actually sure whether we actually want to show the excessive nginx output in either case, what do you think?

Another way of putting my plan above would be to change recovery so that rollback doesn't fail in the first place. Barring that (or maybe in addition), yes, it would be nice to do (1) and (2) only one time each. I also think that if nginx didn't print its identical failure a whole bunch of times, the wall of errors would be less overwhelming, so prettifying (9) could certainly improve things.

Another thought -- how much of an edge case is this whole thing? Do we run into this elsewhere? Would it just be easier to try an nginx restart early on in the code?

**Comment by ohemorange** [comment]
Since `ErrorHandler` is mostly used in rollback and rare scenarios, this is probably not that urgent to resolve.
